cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -0,0 +1,217 @@
+        if (lastMessageId != null) {
+          uri = uri + "?messageId=" + URLEncoder.encode(lastMessageId, "UTF-8");
+        }
+        HttpResponse resp = TetheringUtils.sendHttpRequest(HttpMethod.GET, new URI(peer.getEndpoint())
+          .resolve(uri));
+        switch (resp.getResponseCode()) {
+          case HttpURLConnection.HTTP_OK:
+            handleResponse(resp, peer);
+            break;
+          case HttpURLConnection.HTTP_NOT_FOUND:
+            handleNotFound(peer);
+            break;
+          case HttpURLConnection.HTTP_FORBIDDEN:
+            handleForbidden(peer);
+            break;
+          default:
+            LOG.error("Peer {} returned unexpected error code {} body {}",
+                      peer.getName(), resp.getResponseCode(),
+                      resp.getResponseBodyAsString(StandardCharsets.UTF_8));
+        }
+      } catch (Exception e) {
+        LOG.debug("Failed to create control channel to {}", peer, e);
+      }
+    }
+
+    // Update last message ids in the store for all peers
+    TransactionRunners.run(transactionRunner, context -> {
+      for (Map.Entry<String, String> entry : lastMessageIds.entrySet()) {
+        AppMetadataStore.create(context).persistSubscriberState(entry.getKey(), SUBSCRIBER, entry.getValue());
+      }
+    });
+
+    return connectionInterval;
+  }
+
+  private void handleNotFound(PeerInfo peerInfo) throws IOException {
+    // Update last connection timestamp.
+    store.updatePeerTimestamp(peerInfo.getName());
+
+    TetheringConnectionRequest tetherRequest = new TetheringConnectionRequest(
+      instanceName,
+      peerInfo.getMetadata().getNamespaceAllocations());
+    try {
+      URI endpoint = new URI(peerInfo.getEndpoint()).resolve(TetheringClientHandler.CREATE_TETHER);
+      HttpResponse response = TetheringUtils.sendHttpRequest(HttpMethod.POST,
+                                                             endpoint,
+                                                             GSON.toJson(tetherRequest));
+      if (response.getResponseCode() != 200) {
+        LOG.error("Failed to initiate tether with peer {}, response body: {}, code: {}",
+                  peerInfo.getName(), response.getResponseBody(), response.getResponseCode());
+      }
+    } catch (URISyntaxException | IOException e) {
+      LOG.error("Failed to send tether request to peer {}, endpoint {}",
+                peerInfo.getName(), peerInfo.getEndpoint());
+    }
+  }
+
+  private void initializeMessageIds() {
+    List<String> peers;
+    try {
+      peers = store.getPeers().stream()
+        .map(PeerInfo::getName)
+        .collect(Collectors.toList());
+    } catch (IOException e) {
+      LOG.warn("Failed to get peer information", e);
+      return;
+    }
+
+    TransactionRunners.run(transactionRunner, context -> {
+      for (String peer : peers) {
+        String messageId = AppMetadataStore.create(context).retrieveSubscriberState(peer, SUBSCRIBER);
```

Comment:
Should create the AppMetadataStore outside of the `for` loop.

Reply:
done